src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/Components/ComponentMarkupEncodingPass.cs

```diff
@@ -131,119 +147,71 @@ public override void VisitHtml(HtmlContentIntermediateNode node)
 
             // If we reach here, it means we have successfully decoded all content.
             // Replace all token content with the decoded value.
-            for (var i = 0; i < node.Children.Count; i++)
+            var decodedIndex = 0;
+
+            foreach (var child in node.Children)
             {
-                var child = node.Children[i];
-                if (child is not IntermediateToken token || string.IsNullOrEmpty(token.Content))
+                if (child is not HtmlIntermediateToken token || token.Content.IsNullOrEmpty())
```

Comment:
>if (child is not HtmlIntermediateToken token || token.Content.IsNullOrEmpty()) [](http://example.com/codeflow?start=16&length=78)

might be a bit cleaner if this were looping over a PooledArrayBuilder<(string, HtmlIntermediateToken)>